### DISCUSSION TITLE
BeanFactory와 ApplicationContext의 차이점은 무엇인가?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#160] [Spring Core]
.

**Comment by J-I-H-O** [comment]
**<BeanFactory vs ApplicationContext>**

BeanFactory는 스프링 컨테이너의 최상위 인터페이스로, 빈의 생성과 의존성 주입을 담당합니다. 
이때 BeanFactory는 **Lazy-loading 방식**을 사용합니다. 즉, **빈 로딩 요청을 받는 시점에 on-demand로 인스턴스를 만들고 로딩**합니다.

```java
BeanFactory factory = new XmlBeanFactory(
            new InputStreamResource(
            new FileInputStream("oraclejavacommunity.xml"))); // (1)
OracleJavaComm ojc = (OracleJavaComm)factory.getBean("oracleJavaBean"); // (2)
```

(1)번에서 BeanFactory에 의해 설정 파일이 로드 되더라도 어떠한 빈도 인스턴스화 되지 않습니다.
(2)번에서 getBean() 메서드에 의해 요청을 받는 시점에 인스턴스를 만들고 로드합니다.

---

ApplicationContext 는 BeanFactory의 기능을 상속받아 제공합니다. 
이때 BeanFactory가 제공하는 기능 뿐만 아니라, 애플리케이션 개발을 위한 부가 기능을 제공합니다. 

> - 메시지 소스를 활용한 국제화 기능 : 한국에서 접속하면 한국어로, 영어권에서 접속하면 영어로 출력
> - 환경 변수 : 로컬, 개발, 운영 환경에 따라 구분해서 처리
> - 애플리케이션 이벤트 : 이벤트를 발행하고 구독하는 모델을 편리하게 지원
> - 편리한 리소스 조회 : 파일, 클래스 패스, 외부 등에서 리소스를 편리하게 조회할 수 있도록 함

ApplicationContext는 **Pre-loading 방식**을 사용합니다. **모든 빈들과 설정 파일들이 ApplicationContext에 의해 로드 요청이 될 때 인스턴스로 만들어지고 로드**됩니다.

```java
ApplicationContext context = new ClassPathXmlApplicationContext("oraclejavacommunity.xml"); // (1)
OracleJavaComm ojc = (OracleJavaComm) factory.getBean("oracleJavaBean"); // (2)
```

(1)번에서 모든 빈들이 미리 인스턴스화 됩니다.
(2)번에서 미리 컨테이너에서 만들어진 빈들이 리턴됩니다.

**Comment by le2sky** [comment]
BeanFactory는 기본적인 IoC 컨테이너로 지연 로딩(빈 요청시까지 생성을 미룸)을 이용하며 경량 애플리케이션에 적합할 수 있습니다.
ApplicationContext는 BeanFactory를 확장하고, 즉시 로딩을 이용합니다.
ApplicationContext는 BeanFactory를 확장하면서 추가 기능(국제화, 이벤트 발행, 환경 변수)을 제공합니다.
일반적으로 ApplicationContext가 더 많은 기능을 제공하기 때문에 특별한 이유가 없다면 ApplicationContext를 사용할 것을 권장합니다.
